common/linkutils/link.go

```diff
@@ -0,0 +1,109 @@
+package linkutils
+
+import (
+	"errors"
+	"net/url"
+	"strings"
+)
+
+var ErrInvalidURL = errors.New("URL format is incorrect")
+
+// URL focuses on query params and treats the rest as immutable string.
+// You can use any URL string to construct this object.
+// Its primary goal is to "Expose Query Manipulation".
+// Under the hood it uses url.URL library.
+type URL struct {
+	baseURL     string
+	queryParams url.Values
+	fragment    string
+}
+
+// NewURL will be constructed given valid full url which may have query params.
+func NewURL(base string) (*URL, error) {
+	delegateURL, err := url.Parse(base)
+	if err != nil {
+		return nil, errors.Join(err, ErrInvalidURL)
+	}
+
+	values, err := url.ParseQuery(delegateURL.RawQuery)
+	if err != nil {
+		return nil, errors.Join(err, ErrInvalidURL)
+	}
+
+	result := &URL{
+		queryParams: values,
+		fragment:    delegateURL.Fragment,
+	}
+
+	// Given url.URL structure of:
+	// scheme://[userinfo@]host/path[?query][#fragment]
+	// omit query and fragment to acquire base of URL
+	delegateURL.RawQuery = ""
+	delegateURL.RawFragment = ""
+	delegateURL.Fragment = ""
+	result.baseURL = delegateURL.String()
+
+	return result, nil
+}
+
+func (u *URL) WithQueryParamList(name string, values []string) {
+	u.queryParams[name] = values
+}
+
+func (u *URL) WithQueryParam(name, value string) {
+	u.queryParams[name] = []string{value}
+}
+
+func (u *URL) RemoveQueryParam(name string) {
+	delete(u.queryParams, name)
+}
+
+// ToURL relies on String method.
+func (u *URL) ToURL() (*url.URL, error) {
+	return url.Parse(u.String())
+}
+
+func (u *URL) String() string {
+	return u.baseURL + u.queryValuesToString() + u.fragmentToString()
+}
+
+/*
+Return options:
+
+	=> empty string
+	=> list of query parameters prefixed with `?`
+*/
+func (u *URL) queryValuesToString() string {
+	// We are not fully happy with strict encoding provided by url library
+	// some special symbols are allowed
+	result := u.queryParams.Encode()
+	if len(result) != 0 {
+		for before, after := range map[string]string{
+			"%40": "@",
+			"%24": "$",
+			"%2C": ",",
+		} {
+			result = strings.ReplaceAll(result, before, after)
+		}
```

Comment:
Not sure how I feel about this. On the one hand you're probably right that they're technically allowed. But there's "allowed by the spec" and then there's how common implementations treat them, as well as practical implications. Take the `$` symbol - in the wrong place (e.g. in the middle of a script) it might try to do string interpolation. Is it wrong? No. Is it hazardous? Maybe. It's also a regex anchor symbol, which - again, in the wrong context - might impact the correctness of the output.

Is there a specific reason to do this? Like does some API dislike these escape codes? Unless there's a good reason to do this, my inclination is to say that we should abide by the internal escape logic.

Reply:
Microsoft Dynamics CRM uses special query syntax. Nothing close to "traditional" URL encoding rules.
I think it is better that tests for this are moved back and live with Microsoft. This package may allow exceptions to be registred. I won't impose such limitations for all connectors, indeed thats bad and I don't like it.
Those query parameters if not sanitised could lead to bad outcomes if they haven't put precautions in place. I will treat those as unique "language" per connector.